Add project last update stdout

Glutexo opened the pull request:
Add method to `Project` model that fetches stdout capture of its last update job. It uses `'related'` property of the Project. Returns `nil` if there is no update job for the Project.

Fetching and presenting this information is a workaround for cloning repositories from HTTPS servers with untrusted SSL certificates. Tower API does not offer an option to disable SSL check, so with the stdout user can at least see, why the cloning actually failed.

Alongside fix some subtle bugs in the `Api` class:

* Non-JSON responses were not logged correctly. This bug was hidden because with the default log level `debug` logs are skipped.
* `Faraday` was not available in the `Api` class, resulting it raising `LoadError` while catching `Faraday::*` exception.

https://bugzilla.redhat.com/show_bug.cgi?id=1513616

Required for [#72 in `manageiq-providers-ansible_tower`](https://github.com/ManageIQ/manageiq-providers-ansible_tower/pull/72).

Glutexo commented:
Not sure whom should I ask for review. Maybe @bdunne or @jrafanie?

jrafanie commented on lib/ansible_tower_client/api.rb:
```diff
@@ -1,4 +1,5 @@
 require 'pp'
+require 'faraday'
```
I would put a comment here:
`# We use Faraday exception class constants in method_missing below`

Or, just put the `require 'faraday'` in `method_missing` below since it's the only place (I think) that uses the `Faraday` constant.

jrafanie commented on lib/ansible_tower_client/base_models/project.rb:
```diff
@@ -17,5 +17,13 @@ def update
     end
+
+    def last_update_stdout(format = 'txt')
+      return if !related.respond_to?(:last_update) or related['last_update'].nil?
```
I believe `or` and `||` are the same in this case... Generally, `||` is preferred except in the very specific cases where `or` might be helpful.  I still can't understand `or` most times and stay away from it.  http://www.virtuouscode.com/2010/08/02/using-and-and-or-in-ruby/

jrafanie commented on lib/ansible_tower_client/base_models/project.rb:
```diff
@@ -17,5 +17,13 @@ def update
+      return if !related.respond_to?(:last_update) or related['last_update'].nil?
+
+      last_update_stdout_uri = URI("#{related['last_update']}stdout")
+      last_update_stdout_uri.query = URI.encode_www_form(format: format)
```
Sorry, I was looking at api.rb... It still feels like there should be a URI building method elsewhere.

Glutexo replied:
I got inspired by https://github.com/Glutexo/ansible_tower_client_ruby/blob/afd22258e79804591a630b2ac5a2cd8d23c16c5d/lib/ansible_tower_client/base_models/job.rb#L11, only trying to avoid adding the query as a manually entered string. But yeah, we’re getting a bit un-DRY here. I’ll try to think out some an extraction to some helper class/method.

Glutexo replied:
Actually, the stdout capture can be retrieved from the JSON resource too. Thus there is no need to request `/stdout?format=txt`. We can simply have

```ruby
    def last_update
      return unless related['last_update'].present?
      ProjectUpdate.new(api, api.get(related['last_update']))
    end
```

and avoid the full URI machinery.

Using the same reasoning we can get rid of `Job::stdout` as well as unnecessary. Since this is an API gem, I’d rather open an issue for it instead of blindly removing it.

jrafanie replied:
I'm 👍 on less URI building.  I'm pretty new to this gem.  Does your `last_update` method return the same thing as what you have now?  Finding the most recent `project_updates` reads better to me.

Glutexo replied:
The new `last_update` method returns a whole `ProjectUpdate` object, the original `last_update_stdout` returned only the stdout string. This requires an update on the MIQ provider side too, but that’s no problem.

I can avoid using `related['last_update']` and fetch the most recent update manually by `api.project_updates.all('project' => project.id, 'order_by' => '-id').first`.

Glutexo commented:
I fixed the `require`’s and the `respond_to?`’s. I also simplified the `last_update` method to return the whole object and not only the stdout. This avoided URI building. I stuck to using `related['last_update']` though as they are used quite a lot in this gem and this approach is IMO more readable than building a custom query.

Is it better now, @jrafanie?

jrafanie commented on lib/ansible_tower_client/base_models/project.rb:
```diff
@@ -17,5 +17,10 @@ def update
+    def last_update
+      return if related['last_update'].blank?
+      api.project_updates.find(related['last_update'])
+    end
```
I'm confused.  Is this method called?  I don't see it.

jrafanie replied:
The other changes look good. 👍

Glutexo replied:
It is not called in the API itself. The `manageiq-providers-ansible_tower` will use it.

jrafanie replied:
I see calls to `last_updated` but not `last_update` in the https://github.com/ManageIQ/manageiq-providers-ansible_tower/pull/72 that uses this code change.  Is this right or a typo?  It looks like it's being called on a `project`.

jrafanie replied:
I'm fine with adding new methods but have seen many times where we add code we are sure we'll need but never end up using.  I want to confirm it's actually going to be used next.

Finally, if you're comfortable with `git rebase -i`, it would be nice to squash some of the commits with "fix" or "style" in the name so each commit is standalone with a single type of change.  I don't personally like squashing all commits, but minor typos/style/fix commits are definitely worth squashing.

Glutexo replied:
The `project.last_updated` in [ManageIQ/manageiq-providers-ansible_tower#72](https://github.com/ManageIQ/manageiq-providers-ansible_tower/pull/72) is really a typo. Thanks for finding out! I’ll fix it along with the other changes requested for that PR.

I’ll do the squashing as you suggested.

jrafanie replied:
Don't worry about squashing commits if it becomes difficult.  It's generally good practice to commit "oops/typos/fixes" separate from logical change commits and forcing yourself to try to do this makes it easier to squash commits.  Ping me when it's ready to go.

Glutexo commented:
@jrafanie I squashed the commits. I also bumped the version from `0.13.0` to `0.14.0` so our `manageiq-providers-ansible_tower` “gem” can depend on the new method.

bdunne commented on lib/ansible_tower_client/api.rb:
```diff
@@ -96,12 +96,13 @@ def system_job_templates
     end
 
     def method_missing(method_name, *args, &block)
+      require 'faraday'
```
Why `require 'faraday'` here?

jrafanie replied:
It's for the exception below 👇

bdunne replied:
That won't actually work for `def/rescue/end` because those constants wouldn't exist at method definition time unless faraday is already required somewhere else.  I think it would work only for `def/require/begin/rescue/end/end`.

bdunne commented on spec/factories/responses.rb:
```diff
@@ -26,7 +26,10 @@
     name       { "#{type}-#{id}" }
     url        { "/api/v1/endpoint/" }
-    related    { {"survey_spec" => "example.com/api", 'inventory' => 'inventory link', 'stdout' => 'example.com/api'} }
+    related    { {"survey_spec" => "example.com/api",
```
Avoid multi-line blocks with `{}`.

Glutexo replied:
To be honest, I do not like this using `{}` for one-liners and `do … end` for multi-line blocks much. They behave slightly differently (not in this case) and here it breaks the nice symmetry. I understand the style standards however, so fixed.

bdunne commented on spec/middleware/raise_tower_error_spec.rb:
```diff
@@ -55,6 +57,14 @@
     end
+
+    it "logs error with formatted JSON response body" do
+      response_body = { 'error' => 'missing these attributes' }.pretty_inspect
```
Any reason not to put the string here?

bdunne commented on CHANGELOG.md:
```diff
@@ -4,6 +4,11 @@ All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
+### Added
```
Thanks for updating the changelog!

jrafanie commented:
@Glutexo please mention the people requesting changes if you're done implementing the changes... We don't always notice pushes to your PR branch or know if you're finished.  Thanks!
